rpc: bi-directional streaming for predicate-based channel acceptance

Crypt-iQ opened the pull request:
My attempt at #3005.  I made sure not to have any deadlocks like the `SendPayment` bi-directional stream has.  The funding key, revocation point, and other points probably aren't necessary for the RPC client, but I sent everything from the `*lnwire.OpenChannel` message for the sake of completeness.  Only one client can be active at a time. Added integration tests.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -908,6 +918,85 @@ message SendToRouteRequest {
     Route route = 4;
 }
 
+message ChannelAcceptRequest {
```
I don't think we need to include _all_ of this information. The keys on our side are auto generated, and don't provide any additional context. Minimally, we can leave off all the keying and revocation information.

Crypt-iQ replied:
Done

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+
+// DefaultChannelAcceptor implements the ChannelAcceptor interface
+// and will be used as the default in the funding manager.
+type DefaultChannelAcceptor struct {
```
Is this actually the default? This seems like a special case, RPC controlled variant. 

As far as other default, we may also eventually (in distinct PRs), want to provide a set of acceptors that implement thresholds based off of:
  * mission control state
  * autopilot scoring state (can leverage the scoring system to make decisions off of node connectivity, total channel age, etc)
  * reject nodes we already have channels to
  * etc, etc

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+// sent when requesting to open an inbound channel. This is sent
+// to the RPC streaming client.
+type OpenChannelRequest struct {
+	node        *btcec.PublicKey
```
Public yet the fields are private?

Crypt-iQ replied:
Done

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+// ChannelAcceptor is an interface that represents a predicate on the
+// OpenChannel message and the peer who sent the message.
+type ChannelAcceptor interface {
+	Accept(node *btcec.PublicKey, req *lnwire.OpenChannel) bool
```
Why not use `OpenChannelRequest` here?

Crypt-iQ replied:
Done

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+
+	// If the client is not nil, proxy the arguments to Accept to the RPC
+	// client.
+	if d.client != nil && d.client.active {
```
I think the "default" can be even simpler, and just return "true". We could then create a higher level acceptor (the RPC one) that _embeds_ the default acceptor. Taking a step even further, we could allow _chaining_ acceptors, returning the conjunction of the registered acceptors back to the client.

Crypt-iQ replied:
I think that's a great idea!  It think all the acceptors should be separated though.  I think there are several approaches to this idea of chaining acceptors:

- If the bitwise `&` is not 1, then reject the open channel request.
- Optionally have some sort of "learning" in which a few acceptors return `false` but the majority return `true` but the node still accepts the open channel request.  I'm thinking that over time this model can be improved where there is some total score based on many different factors like you mentioned above.  If the score is above a threshold, accept.
- If the RPC acceptor is specified, this gives "veto" power over all other acceptors.  I think if the RPC acceptor is defined, the other acceptors can be run to send the result and the data used to reach the result back to the client.  This would allow a client to utilize their own score-based approach.

Kind of a brain-dump and maybe quite complicated but I think that would be an ideal direction for these acceptors.  I think it could be complicated for a user to manually specify what their thresholds / scores for various acceptors are and it would make sense for `lnd` to provide this functionality automagically.  For more advanced routing nodes, they could ultimately determine scores.

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+		// Now we wait for a response. If the channel is closed, return false.
+		for {
+			select {
+			case accept, ok := <-d.client.recvChan:
```
Should listen on some sort of quit channel (or context).

Crypt-iQ replied:
Done

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+		for {
+			select {
+			case accept, ok := <-d.client.recvChan:
+				if !ok {
```
IIRc, when a channel is closed, the default value is returned, so unless we want to add additional logic here, we can just return `accept` directly.

Crypt-iQ replied:
Fixed

Roasbeef commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+		// Now we wait for a response. If the channel is closed, return false.
+		for {
+			select {
+			case accept, ok := <-d.client.recvChan:
```
What about time out logic? Do we want to return `false` if our request isn't responded to within period `T`? Would seem that processing time will depend on the client side implementation (can it do concurrent requests, is it FIFO, etc).

Crypt-iQ replied:
I just did a default of 15 seconds, could change this and have this user-specified.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -908,6 +918,85 @@ message SendToRouteRequest {
+}
+
+message ChannelAcceptResponse {
+    /// Whether or not the client accepts the channel.
```
If we add a new request/response ID, then the client's implementation will be able to respond to requests out of order, and also ensure we're able to properly dispatch responses to the correct peer. Atm, the funding mgr is single threaded, but that may change in the future as we add more advanced features like internal channel batching.

Crypt-iQ replied:
Fixed

Roasbeef commented on lnd_test.go:
```diff
@@ -13939,6 +13939,91 @@ func testChannelBackupRestore(net *lntest.NetworkHarness, t *harnessTest) {
+	}()
+}
+
+func testPredicateChannelAccept(net *lntest.NetworkHarness, t *harnessTest) {
```
Integration tests have been moved to a new directory since this PR was created.

halseth commented on channel_acceptor.go:
```diff
@@ -0,0 +1,111 @@
+package lnd
```
possible to move this out of the `lnd` package?

Crypt-iQ replied:
Yup will do

halseth commented:
Needs rebase!

Crypt-iQ commented:
PTAL @halseth @Roasbeef 

I changed a bit of the logic and added a `ChainedAcceptor` which can combine 2 or more different `ChannelAcceptor` implementations.  Let me know what you think!

cfromknecht commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+		return false
+	}
+
+	for {
```
can this ever loop?

Crypt-iQ replied:
Ah no

Crypt-iQ replied:
Ah no

cfromknecht commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+			select {
+			case responseChan <- rpcResp.Accept:
+			default:
+				break
```
`break` is unnecessary

cfromknecht commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+			select {
+			case r.client.SendChan <- chanReq:
+			case <-r.client.closeChan:
+				break
```
similar comment here wrt to `break`

cfromknecht commented on lnrpc/rpc.proto:
```diff
@@ -905,6 +915,58 @@ message SendToRouteRequest {
+    bytes pending_chan_id = 3;
+
+    /// The funding amount in satoshis that initiator wishes to use in the channel.
+    int64 funding_amt = 4;
```
any reason that `int64` was chosen instead of `uint64`? none of these values can be negative

Crypt-iQ replied:
Yup they should be uint64

cfromknecht commented on lntest/itest/lnd_test.go:
```diff
@@ -13928,6 +13928,93 @@ func testHoldInvoicePersistence(net *lntest.NetworkHarness, t *harnessTest) {
+	}
+
+	// Accept the channel request.
+	acceptReponse := &lnrpc.ChannelAcceptResponse{
```
is the pending chanid on all these responses zero?

Crypt-iQ replied:
Nope it's set in the `respondChannelAccept` method

Roasbeef commented on channelacceptor/chainedacceptor.go:
```diff
@@ -0,0 +1,36 @@
+package channelacceptor
```
Could shorten the package name to just: `chanacceptor`.

Roasbeef replied:
Would be nice to have some basic coverage tests for this package now that it no longer lives in the main package.

Roasbeef commented on channelacceptor/interface.go:
```diff
@@ -0,0 +1,21 @@
+// inbound channel. This information is provided to each acceptor so that they
+// can each leverage their own decision-making with this information.
+type OpenChannelRequest struct {
+	Node        *btcec.PublicKey
```
Missing `godoc` comments for these two.

Roasbeef commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+// which includes whether or not they accept the open channel request and to
+// which PendingChannelID the response is to.
+type OpenChannelResponse struct {
+	Accept           bool
```
Missing `godoc` comments here and below.

Roasbeef commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+}
+
+// UnregisterClient unregisters the RPC streaming client.
+func (r *RPCAcceptor) UnregisterClient() {
```
Shouldn't this be on the `ChanAcceptorClient` struct? Otherwise, we can't individually unregister clients.

Roasbeef replied:
Nvm, I see how the `RegisterClient` method works now, no strong preference, but it could also be done with an explicit constructor that instead passes in the `ChanAcceptorClient`.

Roasbeef commented on channelacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,180 @@
+
+		select {
+		case rpcResp := <-r.client.RecvChan:
+			// This is an OpenChannelResponse received from the RPC server. We
```
Would lift the comment here and below to reside right on top of the top-level `select` case.

Roasbeef commented on rpcserver.go:
```diff
@@ -5047,3 +5055,110 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+// time for this RPC.
+func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer) error {
+
+	var acceptorClient *channelacceptor.ChanAcceptorClient
```
Can combine both of these declarations under a single `var` statement:
```go
var (
...
)
```

Roasbeef commented on rpcserver.go:
```diff
@@ -5047,3 +5055,110 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+	openChanPredicate := r.server.fundingMgr.cfg.OpenChannelPredicate
+	chainedAcceptor, ok := openChanPredicate.(*channelacceptor.ChainedAcceptor)
+	if !ok {
+		// The funding manager is not using the ChainedAcceptor.
```
We'll _always_ have the chained acceptor active if this RPC method is also activated?

Crypt-iQ replied:
Since the funding manager accepts a `ChannelAcceptor`, I thought that there could be a case where other implementations besides the `ChainedAcceptor` could also be passed in.  But it seems to me with the current design, `ChainedAcceptor` should be the only implementation that should be passed in since it allows more flexibility.  I can change this and the fundingmanager's `start` and `stop` methods.

Roasbeef commented on lntest/itest/lnd_test.go:
```diff
@@ -13928,6 +13928,93 @@ func testHoldInvoicePersistence(net *lntest.NetworkHarness, t *harnessTest) {
+		lntest.OpenChannelParams{
+			Amt: 100000,
+		},
+	)
```
All channels that end up being created should be closed at the end of the test.

Roasbeef commented on rpcserver.go:
```diff
@@ -5067,3 +5075,108 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+
+	// Locate the RPCAcceptor within the ChainedAcceptor and register this RPC client.
+	openChanPredicate := r.server.fundingMgr.cfg.OpenChannelPredicate
+	chainedAcceptor, _ := openChanPredicate.(*chanacceptor.ChainedAcceptor)
```
Whenever a type assertion is done, the second variable should _always_ be checked. 

It's also a code smell that we're accessing the _config_ of the funding manager here in order to access the RPC acceptor. Atm, the rpc acceptor is a singleton, so we could add a package level method to obtain the latest version of the acceptor.

Roasbeef commented on rpcserver.go:
```diff
@@ -5067,3 +5075,108 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+	openChanPredicate := r.server.fundingMgr.cfg.OpenChannelPredicate
+	chainedAcceptor, _ := openChanPredicate.(*chanacceptor.ChainedAcceptor)
+
+	for _, acceptor := range chainedAcceptor.Acceptors {
```
Similar comment here re just obtaining the global rpc acceptor instance.

Roasbeef commented on rpcserver.go:
```diff
@@ -5067,3 +5075,108 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+
+	for {
+		select {
+		case req := <-acceptorClient.SendChan:
```
These channel attributes (to be morphed into methods based on the comments above), would be better named as response/request rather than send/recv. IMO the intent is more clear with the request/response naming.

Roasbeef commented on fundingmanager.go:
```diff
@@ -487,6 +493,17 @@ func (f *fundingManager) Start() error {
+	for _, acceptor := range chainedAcceptor.Acceptors {
+		rpcAcceptor, ok := acceptor.(*chanacceptor.RPCAcceptor)
+		if ok {
+			rpcAcceptor.Start()
```
Shouldn't the RPC acceptor _already_ be started if it has been passed in at this point?

Roasbeef commented on fundingmanager.go:
```diff
@@ -738,6 +755,17 @@ func (f *fundingManager) Stop() error {
+	for _, acceptor := range chainedAcceptor.Acceptors {
+		rpcAcceptor, ok := acceptor.(*chanacceptor.RPCAcceptor)
+		if ok {
+			rpcAcceptor.Stop()
```
Abstraction leak here, also same comment about it already being stopped (or stopped elsewhere) at this point.

Crypt-iQ commented on chanacceptor/chainedacceptor.go:
```diff
@@ -0,0 +1,36 @@
+	result := true
+
+	for _, acceptor := range c.Acceptors {
+		switch a := acceptor.(type) {
```
I think this will be needed in the future when other acceptors are added.

Crypt-iQ replied:
My bad, you're right.  Should all be abstracted away

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,199 @@
+// client who will respond with the ultimate decision. This assumes a registered
+// client exists.
+func (r *RPCAcceptor) Accept(req *OpenChannelRequest) bool {
+	// Register a new channel for the PendingChannelID associated with this
```
Good point!

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,199 @@
+	return r.client, nil
+}
+
+// UnregisterClient unregisters the RPC streaming client.
```
Will there ever be a case where there is more than one client?  How would that work?  If so, this method can then just accept a `ClientID` and call the `Unregister` method on the client whose `ClientID` matches the argument.

halseth replied:
I assumed more clients would just mean more chained acceptors. So if 4 clients are active, they must all accept for the channel to be accepted. This would allow more general operation, and also some interesting usecases around distributed acceptance.

Crypt-iQ replied:
Yeah like threshold acceptance of channels in a sense.

halseth commented on rpcserver.go:
```diff
@@ -5067,3 +5075,108 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+	chainedAcceptor, _ := openChanPredicate.(*chanacceptor.ChainedAcceptor)
+
+	for _, acceptor := range chainedAcceptor.Acceptors {
+		rpcAcceptor, ok = acceptor.(*chanacceptor.RPCAcceptor)
```
To more easily support multiple RPC clients, and also simplify the registration code a little bit, you could create a `NewRPCAcceptor` method that can be called every time this RPC is called. This acceptor would be strictly tied to this rpc context (so no registration/unregistration would be necessary), and would be added to the current chain of acceptors. When the RPC context is done, it would be removed from the chain.

Crypt-iQ replied:
Yeah I agree with this approach, I started implementing the other way (multiple clients per RPCAcceptor) and found it more difficult to reason about.

halseth commented on fundingmanager.go:
```diff
@@ -487,6 +493,17 @@ func (f *fundingManager) Start() error {
+	// Locate the RPCAcceptor within the ChainedAcceptor and start it if it exists.
+	openChanPredicate := f.cfg.OpenChannelPredicate
+	chainedAcceptor, _ := openChanPredicate.(*chanacceptor.ChainedAcceptor)
+	for _, acceptor := range chainedAcceptor.Acceptors {
```
The fundingmanager shouldn't have to be aware of the individual acceptors.

halseth commented:
ping @Crypt-iQ  :)

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,199 @@
+		default:
+		}
+
+		// If a client does not exist, continue.
```
Don't think this is necessary since now a client is *always* active per @halseth suggestion

Crypt-iQ commented:
PTAL @halseth @Roasbeef @cfromknecht 

Addressed comments - each `RPCAcceptor` now includes its own client by default per @halseth and there is now a `ChainedAcceptor` singleton which has an expanded API than it previously did.  Travis coverage is failing (haven't had time to look at this).  Could have expanded on the `acceptor_test` unit tests now that the usage of the `RPCAcceptor` and `ChainedAcceptor` has changed, but figured it was more prudent to get a quicker review on these changes.

halseth commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,198 @@
+	r.acceptClientsMtx.Unlock()
+
+	select {
+	case r.acceptChan <- req:
```
I find this way of communicating with the RPCserver a bit convulated. 

Not totally sure what would be a better way thought. Would it work having the rpcserver define a closure that would do the demultiplexing? 

```go



NewAcceptor(
    func(req *OpenChannelRequest) bool {
             respChan := make chan
             newRequets <- {req, respChan}

             return <-respChan
    },
)

go func() {
     for {
         resp := stream.Recv()
          resps <- resp
      }
} 

requests := make(map[pendingID]chan response)
for {
     select{
     case req := <-newRequests:
              requests[pendingid] = respChan
             stream.Send(req)
     case resp := <-resps:
              respChan, ok := requests[pendindid]
              if !ok {continue}

             respChan <- resp
             delete(requets, pendingID)
     }

}
```

Crypt-iQ replied:
This would also let us remove `Start` & `Stop` in the `RPCAcceptor`

Crypt-iQ commented on lntest/itest/lnd_test.go:
```diff
@@ -14106,6 +14106,97 @@ func testHoldInvoicePersistence(net *lntest.NetworkHarness, t *harnessTest) {
+		t.Fatalf("failed to reject channel: %v", err)
+	}
+
+	// Timeout the channel request.
```
`DefaultRPCTimeout` = 15 seconds

halseth replied:
That's a bit long for a integration test, since it will just sit here?

Crypt-iQ replied:
I think the only way to configure this and have the same code coverage would be to specify the timeout when calling the `ChannelAcceptor` rpc method.  I do think it gives the user more control over this new functionality, what do you think?

halseth replied:
Yeah, would that work, or would it be hard because of the bidirectional stream? 

I think either way that it should be configurable, as imagine you want to run some fancy algorithm to determine whether to accept the channel, that is super heavy :P

Crypt-iQ replied:
Should be possible w/ the bidi stream, I agree it should be configurable.  Will report back

Crypt-iQ replied:
Since there can only be one input and one output, we can't really specify the timeout unless it's part of the `OpenChannelResponse`, at which point the timeout would apply to the *next* `OpenChannelRequest` message.  I think this approach is confusing.  The timeout could alternatively be specified as a separate rpc or as a command line option when starting lnd.  The downside of the command line option is that it's not mutable unless the user restarts their lnd instance.

Crypt-iQ commented on chanacceptor/chainedacceptor.go:
```diff
@@ -0,0 +1,87 @@
+	// It is possible to add/remove acceptors to/from this ChainedAcceptor while
+	// it's running. It is also possible to have multiple acceptors of the same
+	// type.
+	chainedSingleton *ChainedAcceptor
```
We still need to grab the `ChainedAcceptor` via `GetChainedAcceptor` when adding the `RPCAcceptor` to the acceptors map, so I think it's needed?

halseth replied:
The RPCserver will call `NewRPCAcceptor` to create the acceptor, then add it to the `ChainedAcceptor`, which has been injected from the server

Crypt-iQ replied:
Yeah I guess it would be better to access the `server` from `rpcserver`

Crypt-iQ commented on chanacceptor/chainedacceptor.go:
```diff
@@ -0,0 +1,87 @@
+
+	c.acceptorsMtx.RLock()
+	for _, acceptor := range c.acceptors {
+		result = result && acceptor.Accept(req)
```
I don't disagree but what's the rationale here? If the result will be false anyways, what's the point of calling the remaining acceptors?

halseth replied:
If we have multiple RPC acceptors, they might be used as some kind of notification that channels are being requested.

Crypt-iQ replied:
Yeah that's what I thought, I'll change to `acceptor.Accept(req) && result`

Crypt-iQ commented on lnrpc/rpc.proto:
```diff
@@ -430,6 +430,16 @@ service Lightning {
+    through a single persistent connection. There may only be one client at a
+    time for this RPC.
+    */
+    rpc ChannelAcceptor (stream ChannelAcceptResponse) returns (stream ChannelAcceptRequest);
```
This compiles on my machine but I might be missing something (running `go1.13`)?

halseth replied:
It won't compile if you checkout this exact commit :) But no blocker

Crypt-iQ replied:
Ahhhh, I'll fix it. Good catch!

halseth commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,39 @@
+
+// Accept is a predicate on the OpenChannelRequest which is sent to the RPC
+// client who will respond with the ultimate decision. This assumes an accept
+// closure has been specified during creation.
```
add note that this is part of the interface

halseth commented on rpcserver.go:
```diff
@@ -5065,3 +5077,160 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+
+			req := newRequest.chanReq
+
+			acceptRequestsMtx.Lock()
```
No mutex needed since only accessed from this goroutine :)

Crypt-iQ replied:
Righto!

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,198 @@
+// the ChainedAcceptor singleton.
+var ErrNoChainedAcceptor = errors.New("unable to get chained acceptor")
+
+// OpenChannelResponse is a struct containing the response from the RPC client
```
I went with `OpenChannelResponse` since it mirrors `OpenChannelRequest`, but I'm totally open to changing it.  I think `ChannelAcceptResponse` is a better name but it doesn't like implicitly convey that it's the `OpenChannelRequest`s opposite if you catch my drift.

halseth replied:
Reason I don't want to use `OpenChannelRequest`/`Response` is that we already have `OpenChannelRequest` and `openChanReq` in used in the codebase iirc 😛

halseth commented on rpcserver.go:
```diff
@@ -5065,3 +5065,8 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
 	}
 }
+
+// ChannelAcceptor method stub.
```
👍

halseth commented on rpcserver.go:
```diff
@@ -5066,7 +5078,153 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
-	return nil
+
+	// Locate the ChainedAcceptor within the fundingmanager.
+	openChanPredicate := r.server.fundingMgr.cfg.OpenChannelPredicate
```
instead of acccessing the funding config here, you can store the predicate as a member of the rpc server. Then you don't need the type assertion below either.

Crypt-iQ replied:
Lmk what you think about my approach

halseth commented on lnrpc/rpc.proto:
```diff
@@ -912,6 +921,58 @@ message SendToRouteRequest {
+
+message ChannelAcceptResponse {
+    /// Whether or not the client accepts the channel.
+    bool accept = 1;
```
it could be useful to return a reject reason here, but can be left for future extension.

halseth commented on server.go:
```diff
@@ -896,6 +897,11 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB,
+
+	// Initialize the ChainedAcceptor and add a DefaultAcceptor.
+	chainedAcceptor := chanacceptor.NewChainedAcceptor()
+	chainedAcceptor.AddAcceptor(&chanacceptor.DefaultAcceptor{})
```
I don't think we need the default acceptor. The chained acceptor can just return true as long as there are no added acceptor.

halseth commented on chanacceptor/acceptor_test.go:
```diff
@@ -0,0 +1,155 @@
+		return
+	}
+
+	atomic.AddUint32(ctr, 1)
```
you must add and load value in the same operation

Crypt-iQ commented on server.go:
```diff
@@ -294,7 +295,8 @@ func noiseDial(idPriv *btcec.PrivateKey) func(net.Addr) (net.Conn, error) {
 	privKey *btcec.PrivateKey,
-	chansToRestore walletunlocker.ChannelsToRecover) (*server, error) {
+	chansToRestore walletunlocker.ChannelsToRecover,
+	chanPredicate chanacceptor.ChannelAcceptor) (*server, error) {
```
From what I can tell it is? Compiled on my machine when I checked out this commit

halseth replied:
Ah, not sure why I thought otherwise 😅

Crypt-iQ commented:
My only complaint with this PR is that the timeout isn't currently configurable.  If somebody wanted to make this interact with a mobile app, a user would only have 15 seconds to decide if they want this open channel.  So it only really works with people who have automated scripts ready-to-go.  On the other hand, the funding manager is currently single-threaded so if a user took a long time to accept/reject channel, it may block funding flows for other channels.

halseth commented on chanacceptor/rpcacceptor.go:
```diff
@@ -0,0 +1,41 @@
+// ChannelAcceptResponse is a struct containing the response from the RPC client
+// which includes whether or not they accept the open channel request and to
+// which PendingChannelID the response is to.
+type ChannelAcceptResponse struct {
```
Looks like this is a simple type that now can be defined locally in the rpcserver, where it is used.

halseth commented on rpcserver.go:
```diff
@@ -76,6 +77,10 @@ var (
 	MaxPaymentMSat = maxBtcPaymentMSat
 
+	// defaultAcceptorTimeout is the time after which an RPCAcceptor will time
+	// out and return false if it hasn't yet received a response.
```
add a TODO to make this configurable?

halseth commented on rpcserver.go:
```diff
@@ -5066,7 +5084,163 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+			var pendingID [32]byte
+			copy(pendingID[:], resp.PendingChanId)
+
+			openChanResp := chanacceptor.ChannelAcceptResponse{
```
actually, looks like we can send the `lnrpc` proto on `responses` directly, and remove `ChannelAcceptResponse`

Crypt-iQ replied:
Good point

Roasbeef commented on chanacceptor/chainedacceptor.go:
```diff
@@ -0,0 +1,65 @@
+	for _, acceptor := range c.acceptors {
+		// We call Accept first in case any acceptor (perhaps an RPCAcceptor)
+		// wishes to be notified about ChannelAcceptRequest.
+		result = acceptor.Accept(req) && result
```
Non-blocking style nit: can be
```go
result &= acceptor.Accept(req)
```

halseth replied:
I don't think we can, since that would short cut the call to `Accept`, see discussion https://github.com/lightningnetwork/lnd/pull/3039#discussion_r322605966

halseth commented on rpcserver.go:
```diff
@@ -5052,7 +5072,165 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
+			// Now that we have the response from the RPC client, send it to
+			// the responses chan.
+			select {
+			case responses <- openChanResp:
```
realized we also need to select on quit here, to handle the case where we shut down the acceptor right as the response comes in.

halseth replied:
I think when we have that in this is ready to land!

Crypt-iQ replied:
@halseth donezo. Also thanks for the extensive review on this PR!
